### ISSUE TITLE
[Feature] Support remote storage for visual regression testing

### ISSUE THREAD

**Issue by gskierk** [microsoft/playwright#29227]
Let us know what functionality you'd like to see in Playwright and what your use case is.
Do you think others might benefit from this as well?

It comes from the comment linked below. I didn't find that issue, so I'm creating one by myself. It is about integrating visual regression testing in Playwright with popular stores (e.g., Amazon S3, Azure Blob Storage).

https://github.com/microsoft/playwright/issues/28626#issuecomment-1863431807

**Comment by pavelfeldman** [comment]
Could you elaborate on what exactly you'd like to see?

**Comment by gskierk** [comment]
Currently, the snapshots are saved on the local machine, and the use case I see is for visual regression testing in CI without the need to store the snapshots in a Git repository but rather in the cloud that we have access to (without the need to purchase external visual regression services).

**Comment by pavelfeldman** [comment]
Your request is clear, thanks!

**Comment by joeljeske** [comment]
Similar to https://github.com/microsoft/playwright/issues/28005

**Comment by akellbl4** [comment]
It would be great to see something that can help to compare screenshots with remote storage. Currently it's the only thing that we are missing and it pauses us from using Playwright.

As temporary option I was thinking of a custom matcher that would use the original one and load screenshot before it's called. I assume it can be not the fastest approach and I'd love to see a native solution from Playwright here.

**Comment by scotto-at-niche** [comment]
Ha, I was just coming on here to log that issue and I see someone's already jumped on it. Thanks @gskierk!

We're running into pretty big usability issues with our custom solution; and as we explore going deep on visual testing through a Playwright/Storybook integration, it would be great to have native support for cloud-based visual diff testing (Amazon S3 would be my team's own use case). Not being able to do this natively is hindering buy-in to Playwright from the broader dev group my SDET team supports.

**Comment by zinaelnahel89** [comment]
Same here we also would need this feature as well as a button that would accept or refuse to save the new screenshot if there are changes . Can you tell us when will this approximately be implemented ?

**Comment by JonnyPriceSC** [comment]
@pavelfeldman is there any updates on this would be really great to have the ability to store the expected snapshots remotely? The recommendation for GitHub repos is less than 5GB if we have 1000s of screenshots that are needed to test across different browsers we would really need this ability to unlock scalable regression testing as the repository advised limits will not support this currently.

**Comment by lukeapage** [comment]
Our setup is we run screenshots for a branch and then in the pipeline, make sure one is kicked off or completed for the equivalent commit on master. Then once both complete we run a screenshot comparison build. We keep the screenshots on a shared drive and have a service that keeps the image locations, hashes, state and serves a website to approve branch changes which when approved sets a pr policy to approved.. 
it has the great advantage of not having conflicts to merge but it’s pretty bespoke and also there is a lot of integration work with in our case azure devops.

it would be great if something was built with plugins so you could mix and match storage providers and source code providers.

**Comment by sizzle168** [comment]
Our project would benefit from such a feature as well! Are there any updates on that topic?

**Comment by kyogitq** [comment]
It would be a massive support for our project too if we get this feature within Playwright.

**Comment by sudharsan-selvaraj** [comment]
I was also looking for a similar solution and end up building one. refer to https://github.com/sudharsan-selvaraj/playshot

**What it covers:**

1. **Support for Multiple Cloud Storage Options**: Currently, the library supports uploading screenshots to Amazon S3 and SFTP servers. However, it can be easily adapted to work with other cloud storage providers.
2. **Automatic Screenshot Upload**: Screenshots are automatically uploaded to the remote server if there are test failures, eliminating the need to save them to version control.
3. **Screenshot Updates**: Users have the ability to individually update screenshots when there are changes in the UI.
4. **Flexible Validation**: The library provides methods for validating visual regressions for both entire pages and individual web elements.
5. **Support for Soft Assertions**: Built-in support for soft assertions is included, allowing the test suite to continue running even if a soft assertion fails, without interrupting the execution of subsequent test steps.

**Comment by zinaelnahel** [comment]
@sudharsan-selvaraj  ohhh that's awesome . Just a quick question. We already implemented the deployment to the S3 AWS bucket and it is working fine we want only the screenshot updates part. Did you implement anything that checks the screenshot in the report and if there is a difference to highlight it and ask if we want to save the new screenshot or not ? And if you implemented this how can I use it without the part of the S3 AWS deployment because this is working already fine for us.

**Comment by sudharsan-selvaraj** [comment]
@zinaelnahel Yup, the library has the support for updating the screenshot if there is any change to the UI. You can refer https://github.com/sudharsan-selvaraj/playshot/blob/main/src/matcher.ts#L116 for the implementation details.

**Comment by zinaelnahel** [comment]
@sudharsan-selvaraj wow that's great and is there a way to use this library without this adapter part because I already have the S3 bucket part working and the screenshots as well as the report are deployed there. I just want to import the PlayShotMatcher and use it directly is that possible ? 
import { test as BaseTest, expect } from '@playwright/test';
import { PlayShot } from 'playwright-cloud-visuals';

const playshot = new PlayShot({
  adapter: s3Adapter, // or sshAdapter for SFTP
  remotePathDelimiter: '__screenshots__',
});

const test =
  BaseTest.extend <{ playShot: PlayShotMatcher }>
  {
    playShot: async ({ page }, use, testInfo) => {
      await use(playshot.createMatcher(page, testInfo));
    },
  };

export default test;

**Comment by sudharsan-selvaraj** [comment]
@zinaelnahel   Could you please open a new issue on my repository so that we can continue our discussion there? This will help us keep the current thread focused and organized. Thank you!

**Comment by sputnyq** [comment]
Hey @pavelfeldman. Any news on the feature :)  ?

**Comment by deviantintegral** [comment]
I was originally interested in this feature because I had thought GitHub couldn’t display images in pull requests stored in GitHub LFS. That’s not the case! You get the full image comparison in PRs as well as easy browsing of the repository directly.

I’m sure there are cases where LFS won’t work. But for those who just want to commit screenshots without blowing up your repo size, consider just using LFS.

Given this is just documenting an existing feature, I think it would be a good improvement to call out or even recommend using LFS in the Playwright docs.

**Comment by stephan-noel** [comment]
Did a bit of research on Github LFS, and in the end I think I will try [Git LFS](https://git-lfs.com/) but not [Github LFS](https://docs.github.com/en/repositories/working-with-files/managing-large-files/configuring-git-large-file-storage).  There are many [implementations of Git LFS and most cloud providers should support it](https://github.com/git-lfs/git-lfs/wiki/Implementations) but I'm not sure if you will get the same seamless PR experience though.

The reason for avoiding Github LFS in particular is that I found [a limitation](https://docs.github.com/en/repositories/working-with-files/managing-large-files/removing-files-from-git-large-file-storage#git-lfs-objects-in-your-repository) in which the only way you can remove a file from counting towards your quota is to either delete and re-create you repository or manually contact support.  A [related discussion](https://github.com/orgs/community/discussions/128695) was started.

Hopefully this will be addressed at some point but I haven't found any plans for it.

**Other issues where this limitation is mentioned**: 
https://github.com/git-lfs/git-lfs/issues/4115
https://github.com/git-lfs/git-lfs/discussions/5225#discussioncomment-4428070

**Stack overflow issue mentioning options to work around this with Github LFS**:
https://stackoverflow.com/questions/34579211/how-to-delete-a-file-tracked-by-git-lfs-and-release-the-storage-quota (virtual assistant or [The BFG](https://rtyley.github.io/bfg-repo-cleaner/))

**Comment by gregberge** [comment]
If you’re considering remote storage, it might be a sign that a third-party tool like [Argos](https://argos-ci.com/) could streamline your workflow 👌

**Comment by ShaunMWallace** [comment]
Would love to be able to offload to a remote storage and just include the url to the results as part of GithubAction

**Comment by BenediktHofirek** [comment]
Anything new here?
